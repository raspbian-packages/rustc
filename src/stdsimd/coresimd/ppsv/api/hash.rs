//! Implements `Hash`.

macro_rules! impl_hash {
    ($id:ident, $elem_ty:ident) => {
        impl hash::Hash for $id {
            #[inline]
            fn hash<H: hash::Hasher>(&self, state: &mut H) {
                union A {
                    data: [$elem_ty; $id::lanes()],
                    vec: $id
                }
                unsafe {
                    A { vec: *self }.data.hash(state)
                }
            }
        }
    }
}

#[cfg(test)]
#[macro_export]
macro_rules! test_hash {
    ($id:ident, $elem_ty:ident) => {
        #[test]
        fn hash() {
            use ::coresimd::simd::$id;
            use ::std::collections::hash_map::DefaultHasher;
            use ::std::hash::{Hash, Hasher};
            use ::std::{mem, clone};
            use clone::Clone;
            type A = [$elem_ty; $id::lanes()];
            let a: A = [42 as $elem_ty; $id::lanes()];
            assert!(mem::size_of::<A>() == mem::size_of::<$id>());
            let mut a_hash = DefaultHasher::new();
            let mut v_hash = a_hash.clone();
            a.hash(&mut a_hash);

            let v = $id::splat(42 as $elem_ty);
            v.hash(&mut v_hash);
            assert_eq!(a_hash.finish(), v_hash.finish());
        }
    }
}
