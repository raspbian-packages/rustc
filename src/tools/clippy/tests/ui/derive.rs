#![feature(plugin)]
#![plugin(clippy)]

#![feature(untagged_unions)]

#![allow(dead_code)]

use std::hash::{Hash, Hasher};

#[derive(PartialEq, Hash)]
struct Foo;

impl PartialEq<u64> for Foo {
    fn eq(&self, _: &u64) -> bool { true }
}

#[derive(Hash)]
struct Bar;

impl PartialEq for Bar {
    fn eq(&self, _: &Bar) -> bool { true }
}

#[derive(Hash)]
struct Baz;

impl PartialEq<Baz> for Baz {
    fn eq(&self, _: &Baz) -> bool { true }
}

#[derive(PartialEq)]
struct Bah;

impl Hash for Bah {
    fn hash<H: Hasher>(&self, _: &mut H) {}
}

#[derive(Copy)]
struct Qux;

impl Clone for Qux {
    fn clone(&self) -> Self { Qux }
}

// looks like unions don't support deriving Clone for now
#[derive(Copy)]
union Union {
    a: u8,
}

impl Clone for Union {
    fn clone(&self) -> Self {
        Union {
            a: 42,
        }
    }
}

// See #666
#[derive(Copy)]
struct Lt<'a> {
    a: &'a u8,
}

impl<'a> Clone for Lt<'a> {
    fn clone(&self) -> Self { unimplemented!() }
}

// Ok, `Clone` cannot be derived because of the big array
#[derive(Copy)]
struct BigArray {
    a: [u8; 65],
}

impl Clone for BigArray {
    fn clone(&self) -> Self { unimplemented!() }
}

// Ok, function pointers are not always Clone
#[derive(Copy)]
struct FnPtr {
    a: fn() -> !,
}

impl Clone for FnPtr {
    fn clone(&self) -> Self { unimplemented!() }
}

// Ok, generics
#[derive(Copy)]
struct Generic<T> {
    a: T,
}

impl<T> Clone for Generic<T> {
    fn clone(&self) -> Self { unimplemented!() }
}

fn main() {}
