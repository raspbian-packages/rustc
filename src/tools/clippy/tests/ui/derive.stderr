error: you are deriving `Hash` but have implemented `PartialEq` explicitly
  --> $DIR/derive.rs:17:10
   |
17 | #[derive(Hash)]
   |          ^^^^
   |
   = note: #[deny(derive_hash_xor_eq)] on by default
note: `PartialEq` implemented here
  --> $DIR/derive.rs:20:1
   |
20 | / impl PartialEq for Bar {
21 | |     fn eq(&self, _: &Bar) -> bool { true }
22 | | }
   | |_^

error: you are deriving `Hash` but have implemented `PartialEq` explicitly
  --> $DIR/derive.rs:24:10
   |
24 | #[derive(Hash)]
   |          ^^^^
   |
note: `PartialEq` implemented here
  --> $DIR/derive.rs:27:1
   |
27 | / impl PartialEq<Baz> for Baz {
28 | |     fn eq(&self, _: &Baz) -> bool { true }
29 | | }
   | |_^

error: you are implementing `Hash` explicitly but have derived `PartialEq`
  --> $DIR/derive.rs:34:1
   |
34 | / impl Hash for Bah {
35 | |     fn hash<H: Hasher>(&self, _: &mut H) {}
36 | | }
   | |_^
   |
note: `PartialEq` implemented here
  --> $DIR/derive.rs:31:10
   |
31 | #[derive(PartialEq)]
   |          ^^^^^^^^^

error: you are implementing `Clone` explicitly on a `Copy` type
  --> $DIR/derive.rs:41:1
   |
41 | / impl Clone for Qux {
42 | |     fn clone(&self) -> Self { Qux }
43 | | }
   | |_^
   |
   = note: `-D expl-impl-clone-on-copy` implied by `-D warnings`
note: consider deriving `Clone` or removing `Copy`
  --> $DIR/derive.rs:41:1
   |
41 | / impl Clone for Qux {
42 | |     fn clone(&self) -> Self { Qux }
43 | | }
   | |_^

error: you are implementing `Clone` explicitly on a `Copy` type
  --> $DIR/derive.rs:65:1
   |
65 | / impl<'a> Clone for Lt<'a> {
66 | |     fn clone(&self) -> Self { unimplemented!() }
67 | | }
   | |_^
   |
note: consider deriving `Clone` or removing `Copy`
  --> $DIR/derive.rs:65:1
   |
65 | / impl<'a> Clone for Lt<'a> {
66 | |     fn clone(&self) -> Self { unimplemented!() }
67 | | }
   | |_^

error: you are implementing `Clone` explicitly on a `Copy` type
  --> $DIR/derive.rs:75:1
   |
75 | / impl Clone for BigArray {
76 | |     fn clone(&self) -> Self { unimplemented!() }
77 | | }
   | |_^
   |
note: consider deriving `Clone` or removing `Copy`
  --> $DIR/derive.rs:75:1
   |
75 | / impl Clone for BigArray {
76 | |     fn clone(&self) -> Self { unimplemented!() }
77 | | }
   | |_^

error: you are implementing `Clone` explicitly on a `Copy` type
  --> $DIR/derive.rs:85:1
   |
85 | / impl Clone for FnPtr {
86 | |     fn clone(&self) -> Self { unimplemented!() }
87 | | }
   | |_^
   |
note: consider deriving `Clone` or removing `Copy`
  --> $DIR/derive.rs:85:1
   |
85 | / impl Clone for FnPtr {
86 | |     fn clone(&self) -> Self { unimplemented!() }
87 | | }
   | |_^

error: aborting due to 7 previous errors

